Jshooks [DO NOT MERGE]

RichardAH opened the pull request:
<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.
-->

## High Level Overview of Change

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

<!--
Please check [x] relevant options, delete irrelevant ones.

* If there is any impact to the public API methods (HTTP / WebSocket), please update https://github.com/xrplf/rippled/blob/develop/API-CHANGELOG.md
  * Update API-CHANGELOG.md and add the change directly in this PR by pushing to your PR branch.
* libxrpl: See https://github.com/XRPLF/rippled/blob/develop/docs/build/depend.md
* Peer Protocol: See https://xrpl.org/peer-protocol.html
-->

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.
-->

<!--
## Test Plan
If helpful, please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
This section may not be needed if your change includes thoroughly commented unit tests.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

RichardAH committed:
quickjs running inside xahaud

RichardAH committed:
more progress on jsvm

RichardAH committed:
compiling

RichardAH committed:
bug fixes

RichardAH committed:
debug

RichardAH committed:
first js hookapi working (accept)

RichardAH committed:
hookjs rollback

RichardAH committed:
otxnid, trace apis for js hooks

RichardAH committed:
js util_raddr

RichardAH committed:
js util_accid

RichardAH committed:
js util_verify

RichardAH committed:
js util_sha512h

RichardAH committed:
bug fix, debugging print

RichardAH committed:
start of js util_keylet

RichardAH committed:
js util_keylet untested compiling

RichardAH committed:
js otxn_type

RichardAH committed:
js otxn_param otxn_slot

RichardAH committed:
js otxn_field otxn_generation otxn_burden etxn_burden

RichardAH committed:
js etxn_fee_base/details/reserve/generation/nonce hook_account

RichardAH committed:
js hook_hash, hook_again, fee_base, ledger_seq, ledger_last_hash, ledger_last_time, ledger_nonce, ledger_keylet

RichardAH committed:
js hook_param hook_param_set hook_skip hook_pos

RichardAH committed:
js state, state_foreign, state_set, state_foreign_set, compiling untested

RichardAH committed:
js slot slot_clear slot_count slot_set slot_size slot_subarray slot_subfield slot_type slot_float compiling not tested

RichardAH committed:
enable js slot apis

dangell7 committed:
fix quickjs build

dangell7 committed:
Update quickjs.cmake

RichardAH committed:
js sto_validate sto_subfield sto_subarray sto_emplace sto_erase, compiling not tested

RichardAH committed:
js float_set float_multiply float_mulratio float_negate float_compare float_sum float_sto float_sto_set float_invert float_divide float_one float_mantissa float_sign float_int float_log float_root

RichardAH committed:
js emit

RichardAH committed:
js prepare otxn_json, compiling not tested

RichardAH committed:
js prepare/emit working

dangell7 committed:
add hso w/ api version

dangell7 committed:
add set js hook tests

dangell7 committed:
clean up test code

dangell7 committed:
[fold] tmp use `hsov1`

Will remove later just did this to avoid to many updates

dangell7 committed:
more js hook tests

RichardAH committed:
js slot_json sto_to_json sto_from_json meta_slot xpop_slot

RichardAH committed:
fix js otxn_id hook_account bugs

RichardAH committed:
fix account type on js otxn_field

dangell7 committed:
fix float_set

dangell7 committed:
fix_float

dangell7 committed:
[fold] denis error

dangell7 committed:
dump

dangell7 committed:
Revert "dump"

This reverts commit b650d72f0b5c6fec17804e8b365b4b251be9cb1c.

dangell7 committed:
fix `sto_emplace`

dangell7 committed:
nit: reorder fields

WietseWind committed:
Merge pull request #326 from Xahau/dev

CI on `jshooks` branch (#317)

WietseWind committed:
CI Split jobs with prev job dependency & CI on `jshooks` (#320) (#328)

* CI on `jshooks` branch

* CI Split jobs with prev job dependency

* No multi branch worker in parallel

---------

Co-authored-by: Denis Angell <dangell@transia.co>

dangell7 committed:
add tests

WietseWind committed:
Fix account_tx sluggishness as per https://github.com/XRPLF/rippled/commit/2e9261cb (#330)

WietseWind committed:
Fix JS float_int return JS instead of JSXFL

dangell7 committed:
cont. tests

dangell7 committed:
cont. tests

dangell7 committed:
remove dummy arg causes `INVALID_ARGUMENT` for keylets without args

dangell7 committed:
fix `sto_subarray`

RichardAH committed:
make accept/rollback actually exit the jsvm

RichardAH committed:
remove promise, date and proxy from the jsvm

RichardAH committed:
first iteration of adding an instruction limit to jshooks

RichardAH committed:
qjs modules, compiling not tested

dangell7 committed:
patch `state_foreign_set ` delete

dangell7 committed:
comment out debug

dangell7 committed:
more testing

RichardAH committed:
TextEncoder/Decoder working in jshooks

RichardAH committed:
add error logging for qjs

dangell7 committed:
add tests

RichardAH committed:
better exit handling accept/rollback/exception/instruction limit

dangell7 committed:
fix v1 errors

dangell7 committed:
add feature

dangell7 committed:
add proper error handling

dangell7 committed:
update tests

RichardAH committed:
rename HooksV1 feature to JSHooks feature

WietseWind committed:
Workaround CentOS7 EOL

dangell7 committed:
fix slot_float return

dangell7 committed:
`float_sign` should be int not bigint

dangell7 committed:
fix return error

dangell7 committed:
build for mac

dangell7 committed:
Merge branch 'dev' into jshooks

tequdev committed:
fix: build error `no type named 'Reader' in namespace 'Json'`

tequdev committed:
fix: not to use `std::lexicographical_compare_three_way`

dangell7 committed:
[fold] preprocessor macros

RichardAH committed:
Merge branch 'dev' into jshooks

RichardAH committed:
Merge branch 'jshooks' into fix-jshooks-at-macos-build

RichardAH committed:
Merge pull request #386 from tequdev/fix-jshooks-at-macos-build

Co-authored-by: Denis Angell <dangell@transia.co>

tequdev commented on src/ripple/app/hook/impl/applyHook.cpp:
```diff
@@ -5481,25 +8256,55 @@ float_divide_internal(int64_t float1, int64_t float2, bool hasFix)
+    if (any_missing(f1, f2))
+        returnJS(INVALID_ARGUMENT);
+
+    int64_t const out = float_divide_internal(*f1, *f2);
```
```suggestion
    bool const hasFix = view.rules().enabled(fixFloatDivide);
    int64_t const out = float_divide_internal(*f1, *f2, hasFix);
```

tequdev commented on src/ripple/app/hook/impl/applyHook.cpp:
```diff
@@ -5510,33 +8315,104 @@ DEFINE_HOOK_FUNCTION(int64_t, float_invert, int64_t float1)
+    if (*f1 == float_one_internal)
+        returnJSXFL(float_one_internal);
+
+    int64_t const out = float_divide_internal(float_one_internal, *f1);
```
```suggestion
    bool const hasFix = view.rules().enabled(fixFloatDivide);
    int64_t const out = float_divide_internal(float_one_internal, *f1, hasFix);
```

tequdev commented on src/ripple/basics/base_uint.h:
```diff
@@ -570,7 +570,7 @@ operator<=>(base_uint<Bits, Tag> const& lhs, base_uint<Bits, Tag> const& rhs)
                                         : std::strong_ordering::less;
 }
-#elif
+
```
bad merge?
`#elif` should be exist

tequdev committed:
fix jshooks build error (#410)

tequdev commented on src/ripple/app/hook/impl/applyHook.cpp:
```diff
@@ -5398,7 +8137,43 @@ DEFINE_HOOK_FUNCTION(
+    if (!sto.has_value())
+        returnJS(INVALID_ARGUMENT);
+
+    returnJS(__float_sto_set(hookCtx, applyCtx, j, sto->data(), sto->size()));
```
`float_sto_set` should return successful value as bigint(XFL)

```suggestion
    int64_t const out = __float_sto_set(hookCtx, applyCtx, j, sto->data(), sto->size());
    if (out < 0)
        returnJS(out);

    returnJSXFL(out);
```

tequdev commented on src/ripple/app/hook/applyHook.h:
```diff
@@ -888,12 +1288,396 @@ class HookExecutor
+
+        ~QuickJSVM()
+        {
+            if (rt)
+                JS_FreeRuntime(rt);
+            if (ctx)
+                JS_FreeContext(ctx);
+            rt = NULL;
+            ctx = NULL;
```
In my environment, Object Leaks error occurred when running a local build, but changing it as suggested resolved the error.

Also, in the [quickjs sample code](https://github.com/bellard/quickjs/blob/6e2e68fd0896957f92eb6c242a2e048c1ef3cae0/fuzz/fuzz_eval.c#L46-L47), JS_FreeRuntime is called after JS_FreeContext, as suggested.

```suggestion
            if (ctx)
                JS_FreeContext(ctx);
            if (rt)
                JS_FreeRuntime(rt);
            ctx = NULL;
            rt = NULL;
```

tequdev commented on src/ripple/app/hook/applyHook.h:
```diff
@@ -888,12 +1288,396 @@ class HookExecutor
+
+        char expr[256];
+
+        int expr_len = 
+            snprintf(expr, 256, "%s(%d)",
+                callback ? "Callback" : "Hook",
+                hookArgument);
+
+        if (expr_len < 7 || expr_len == 256)
```
This change makes it possible to do #332 Request 2.

```suggestion
        int expr_len = 
            snprintf(expr, 256, "%s(%d, otxn_json())",
                callback ? "Callback" : "Hook",
                hookArgument);

        if (expr_len < 20 || expr_len == 256)
```

tequdev commented on src/ripple/app/hook/impl/applyHook.cpp:
```diff
@@ -3640,28 +5573,33 @@ DEFINE_HOOK_FUNCTION(
-    uint32_t write_len)
+DEFINE_JS_FUNCNARG(
+    JSValue,
+    hook_account);
```
Unnecessary semicolon
```suggestion
    hook_account)
```

tequdev replied:
https://github.com/Xahau/xahaud/pull/481

tequdev committed:
JSHooks Instruction Count (#423)

dangell7 committed:
conan

dangell7 committed:
sync

dangell7 committed:
[patch] remit transfer rate

dangell7 committed:
[release] add release build

dangell7 committed:
[fold] update workflow

dangell7 committed:
[fold] fix bad assert

dangell7 committed:
[fold] upgrade artifact@v4

dangell7 committed:
[fold] add external snappy

dangell7 committed:
[fold] remove conan patch

dangell7 committed:
[fold] use rippleci docker image

dangell7 committed:
[fold] update conan

dangell7 committed:
[fold] clang-format

dangell7 committed:
[fold] fix test

dangell7 committed:
fix misc warnings

dangell7 committed:
include optional

dangell7 committed:
fix macos github action

dangell7 committed:
fix misc warnings

dangell7 committed:
add optional include

dangell7 committed:
clang-format

dangell7 committed:
fix bad commit

dangell7 committed:
fix failing test

dangell7 committed:
Update macos.yml

dangell7 committed:
Merge branch 'dev' into tmp-conan

dangell7 committed:
[temp] remove negative rate test

sublimator commented:
I don't know about Clang 14, but for Clang 16, I needed to merge in the dev branch which removes some features that were deprecated in C++17/

dangell7 committed:
Merge branch 'dev' into jshooks

tequdev committed:
fix JSHooks tests (#446)

tequdev committed:
Add bytecode validity check on SetHook (#439)

tequdev committed:
fix sto_from_json error (#438)

tequdev committed:
add replacer for Stringify bigint value (#437)

dangell7 committed:
Merge branch 'dev' into jshooks

dangell7 committed:
[fold] clang-format

dangell7 committed:
reorder free context

Co-authored-by: tequ <git@tequ.dev>

dangell7 committed:
`float_sto_set` return bigint

Co-authored-by: tequ <git@tequ.dev>

dangell7 committed:
[fold] clang-format

dangell7 committed:
[fold] update test wasm

dangell7 committed:
[revert] typo

sublimator committed:
feat(quickjs): add test_sto_emplace (#448)

dangell7 committed:
Merge branch 'dev' into tmp-conan

tequdev commented on src/ripple/app/hook/impl/applyHook.cpp:
```diff
@@ -5511,33 +8254,96 @@ DEFINE_HOOK_FUNCTION(int64_t, float_invert, int64_t float1)
+    RETURNJS_IF_INVALID_FLOAT(*f1);
+
+    if (*f1 == 0)
+        returnJS(0);
```
## float_mantissa

The success value should be returned by XFL(bigint).

```suggestion
        returnJSXFL(0);
```

tequdev replied:
https://github.com/Xahau/xahaud/pull/481

tequdev commented on src/ripple/app/hook/impl/applyHook.cpp:
```diff
@@ -4842,7 +7261,35 @@ DEFINE_HOOK_FUNCTION(int64_t, float_set, int32_t exp, int64_t mantissa)
+    int32_t exp = *e;
+
+    if (mantissa == 0)
+        returnJS(0);
```
## float_set

The success value should be returned by XFL.

```suggestion
        returnJSXFL(0);
```

tequdev replied:
https://github.com/Xahau/xahaud/pull/481

tequdev commented on src/ripple/app/hook/impl/applyHook.cpp:
```diff
@@ -4988,33 +7500,94 @@ DEFINE_HOOK_FUNCTION(
+        returnJS(INVALID_ARGUMENT);
+
+    if (*float1 == 0)
+        returnJS(0);
```
## float_negate

The success value should be returned by XFL.

```suggestion
        returnJSXFL(0);
```

tequdev replied:
https://github.com/Xahau/xahaud/pull/481

dangell7 committed:
Merge branch 'dev' into tmp-conan

dangell7 committed:
[fix] release build issue

tequdev commented on include/xrpl/hook/Enum.h:
```diff
@@ -340,6 +343,14 @@ enum ExitType : uint8_t {
     WASM_ERROR = 1,
     ROLLBACK = 2,
     ACCEPT = 3,
+    JSVM_ERROR = 4,
+    LEDGER_ERROR =
+        5,  // if the ledger contained for example a nonsense hookapi number
```
If the Instruction Limit is reached, the ExitType is currently set to UNSET, but it may be worth adding a new ExitType(`INSTRUCTION_LIMIT_REACHED`) for this case.

tequdev committed:
add tests for `float_sto`, `float_sto_set`, `float_sum` (#462)

tequdev committed:
[JSHooks] Add sto tests (#460)

tequdev committed:
Enable JS Date on using parent ledger close time

tequdev reviewed (commented):
https://github.com/Xahau/xahaud/blob/jshooks/src/quickjs/quickjs.c#L43359

Either `Math.random()` should not be used, or the implementation should be changed to give deterministic results.

sublimator committed:
Update jshooks header

sublimator committed:
Merge remote-tracking branch 'origin/dev' into jshooks

sublimator committed:
Fix validate exception handling #473

sublimator committed:
Add missing JS_FreeValue calls for early returns #468

sublimator committed:
Free out array in slot before returning special case int64 #467

sublimator committed:
Return DOESNT_EXIST in JSHooks sto_erase #463

tequdev committed:
Updated the jshook build script so that it works on macOS and Ubuntu (#464)

sublimator committed:
Bump numFeatures

tequdev committed:
Add tests for `fee_base`, `ledger_keylet`, `meta_slot` (#469)

* add tests for fee_base, ledger_keylet, meta_slot

* clang-format

* clang-format

* update SetJSHook_wasm.h

sublimator committed:
Automatically add MacOS gsed/ggrep paths to PATH if exist

tequdev committed:
Add `xpop_slot` test (#476)

* add xpop_slot test

* clang-format

sublimator committed:
Test for NO_FREE_SLOTS

sublimator committed:
Use official shell script

dangell7 committed:
update conan to 1.86

dangell7 committed:
Merge branch 'dev' into tmp-conan

sublimator committed:
Merge remote-tracking branch 'origin/tmp-conan' into jshooks

sublimator committed:
[CONAN] Run MacOS builds on GH runners

sublimator committed:
[CONAN] Do not build cobalt

sublimator committed:
[CONAN] Use self-hosted runner again

sublimator committed:
[CONAN] check clang --version

sublimator committed:
[CONAN] remove DBOOST_ASIO_DISABLE_CONCEPTS define

sublimator committed:
[CONAN] clear cache

sublimator committed:
[CONAN] clear the tools.build:cxxflags

sublimator committed:
[CONAN] clear the tools.build:cxxflags

sublimator committed:
[CONAN] clean up conf experiments

dangell7 committed:
remove `cxxflags` for macos action

sublimator committed:
[CONAN] simplify matrix

sublimator committed:
[CONAN] fix typo

sublimator committed:
[CONAN] simplify matrix

sublimator committed:
[CONAN] use macos-15

sublimator committed:
[CONAN] use simpler GA runners with conan cache

sublimator committed:
[CONAN] use simpler GA runners with conan cache

sublimator committed:
[CONAN] use simpler GA runners with conan cache

sublimator committed:
[CONAN] use v4 -artifacts

sublimator committed:
[CONAN] remove redundant step

sublimator committed:
[CONAN] fix typo and redundant deps step

tequdev committed:
add Emit tests (#478)

* add tests for emits

* clang-format

* fix otxn_burden

* remove std:cout

* fix tests

sublimator committed:
[CONAN] install nproc

sublimator committed:
[CONAN] no access to 8 core runner

sublimator committed:
[CONAN] manually handle cache saving step

sublimator committed:
Merge branch 'tequdev/jshooks-date' into jshooks

sublimator committed:
[CONAN] manually handle cache saving step in MacOS

sublimator committed:
Mark header file as generated

sublimator committed:
[CONAN] cache build directory too

sublimator committed:
[CONAN] cache build directory too

sublimator committed:
[CONAN] try using abs path

tequdev committed:
fix zero XFL (#481)

sublimator committed:
[CONAN] try using glob

dangell7 committed:
[GA] conan mac os build

dangell7 committed:
[GA] update conan profile

dangell7 committed:
[GA] remove unused workflows

sublimator committed:
[CONAN] cache ccache folder

dangell7 committed:
[GA] tmp disable build cache

dangell7 committed:
[GA] revert

sublimator committed:
[CONAN] fix MacOS ccache folder

sublimator committed:
[CONAN] include-hidden-folders: true

dangell7 committed:
[GA] Remove build caching

sublimator committed:
[CONAN] dummy tests for now

sublimator committed:
[CONAN] fallback cache key with run_id

sublimator committed:
[CONAN] use dummy jobs

sublimator committed:
[CONAN] prime .ccache

sublimator committed:
[CONAN] stable fallback cache key

sublimator committed:
[CONAN] use step conclusion

sublimator committed:
[CONAN] use always() and step conclusion

sublimator committed:
[CONAN] better ccache keys

sublimator committed:
feat: add RUN_TARGET_PATTERN macro with --unittest-arg regex

sublimator committed:
[CONAN] clean up comments

sublimator committed:
[CONAN] Move caching steps into build/dependencies actions

sublimator committed:
[CONAN] Fix Claude's mess

sublimator committed:
[CONAN] Remove straggling .build_deps mention

sublimator committed:
Add :log-test-names: target for --unittest-arg

sublimator committed:
[CONAN] Namespace xahau actions/workflows for easier upstream

sublimator committed:
[CONAN] Do not build on sh runner for PRs targeting dev

sublimator committed:
[CONAN] GA only needs to declare push with ignore-branches

sublimator committed:
Use GA runners with conan/ccache caches (#482)

dangell7 committed:
Update xahau-sh-build-in-docker.yml

dangell7 committed:
refactor checkpatterns remove comments

dangell7 committed:
change `on` fields

dangell7 committed:
fix checkpatterns

dangell7 committed:
tmp fix docker release builder

dangell7 committed:
fix build dir

dangell7 committed:
attempt fix builder

dangell7 committed:
optional release builder

dangell7 committed:
revert some cmake builds

dangell7 committed:
revert docker dependency cache

dangell7 committed:
remove `EXIT_IF_CONTAINER_RUNNING`

dangell7 committed:
test workflow

dangell7 committed:
Update release-builder.sh

dangell7 committed:
add caching

dangell7 committed:
Update xahau-sh-build-in-docker.yml

dangell7 committed:
Update xahau-sh-build-in-docker.yml

dangell7 committed:
eof

dangell7 committed:
Merge branch 'tmp-conan' into jshooks

sublimator committed:
Merge remote-tracking branch 'origin/dev' into jshooks

sublimator committed:
Fix missing includes in Catalogue.cpp for non-unity builds

sublimator committed:
[CONAN] Disable not working static containers

sublimator committed:
Merge remote-tracking branch 'origin/dev' into jshooks

sublimator commented on CMakeLists.txt:
```diff
@@ -7,17 +10,18 @@ if(POLICY CMP0077)
-set(CMAKE_CXX_EXTENSIONS OFF)
-set(CMAKE_CXX_STANDARD 20)
-set(CMAKE_CXX_STANDARD_REQUIRED ON)
+if(POLICY CMP0144)
```
TODO: Fix double CMP0144

sublimator committed:
chore: remove duplicate cmake policy directive from bad merge

gregtatcam committed:
fix: Add AMMv1_3 amendment (#5203)

* Add AMM bid/create/deposit/swap/withdraw/vote invariants:
  - Deposit, Withdrawal invariants: `sqrt(asset1Balance * asset2Balance) >= LPTokens`.
  - Bid: `sqrt(asset1Balance * asset2Balance) > LPTokens` and the pool balances don't change.
  - Create: `sqrt(asset1Balance * assetBalance2) == LPTokens`.
  - Swap: `asset1BalanceAfter * asset2BalanceAfter >= asset1BalanceBefore * asset2BalanceBefore`
     and `LPTokens` don't change.
  - Vote: `LPTokens` and pool balances don't change.
  - All AMM and swap transactions: amounts and tokens are greater than zero, except on withdrawal if all tokens
    are withdrawn.
* Add AMM deposit and withdraw rounding to ensure AMM invariant:
  - On deposit, tokens out are rounded downward and deposit amount is rounded upward.
  - On withdrawal, tokens in are rounded upward and withdrawal amount is rounded downward.
* Add Order Book Offer invariant to verify consumed amounts. Consumed amounts are less than the offer.
* Fix Bid validation. `AuthAccount` can't have duplicate accounts or the submitter account.

yinyiqian1 committed:
fixAMMClawbackRounding: adjust last holder's LPToken balance (#5513)

Due to rounding, the LPTokenBalance of the last LP might not match the LP's trustline balance. This was fixed for `AMMWithdraw` in `fixAMMv1_1` by adjusting the LPTokenBalance to be the same as the trustline balance. Since `AMMClawback` is also performing a withdrawal, we need to adjust LPTokenBalance as well in `AMMClawback.`

This change includes:
1. Refactored `verifyAndAdjustLPTokenBalance` function in `AMMUtils`, which both`AMMWithdraw` and `AMMClawback` call to adjust LPTokenBalance.
2. Added the unit test `testLastHolderLPTokenBalance` to test the scenario.
3. Modify the existing unit tests for `fixAMMClawbackRounding`.

tequdev committed:
Merge branch 'dev' into jshooks

tequdev committed:
Fix `FromJSIntArrayOrHexString` `[]` or `”"` return vector(size=0), not nullptr. (#553)

tequdev committed:
Improve validation for `state_foreign_set` (#554)

* Fix `FromJSIntArrayOrHexString` `[]` or `”"` return vector(size=0), not nullptr.

* improve validation for `state_foreign_set`

tequdev committed:
Fix JSHooks Fee validation (#550)

tequdev committed:
Fix `hook_param_set` to support deletion (#551)

tequdev committed:
add checks the length of currency and issuer for `float_sto` (#518)

fix #461

sublimator committed:
merge: origin/dev into jshooks, accepting upstream changes

sublimator committed:
fix: gcc 13 compile error, per #557 (#564)

tequdev committed:
Fix `slot` to return `bigint` when specified true for second argument (#573)

tequdev committed:
Allow update JSHook Fee on Install/Update (#577)

* Allow update JSHook Fee on Install/Update, handle Instruction Limit Reach properly

* use `JS_FEE_OUT_OF_RANGE` as hook_log_code  instead `JS_FEE_TOO_HIGH`

tequdev committed:
Merge fixAMMv1_3 amendment into featureAMM amendment

tequdev committed:
Merge fixAMMClawbackRounding amendment into featureAMMClawback amendment

RichardAH committed:
Merge pull request #692 from Xahau/sync-2.4.0-rebased

Sync: Ripple(d) 2.4.0

tequdev committed:
Merge commit 'd18c4d9b33ad42c268ab1e97606da6368e89587d' into jshooks

tequdev committed:
Merge branch 'dev' into jshooks

tequdev committed:
fix clang-format workflow to exclude quickjs files

tequdev committed:
update generate_extern.sh

tequdev committed:
update ordering.txt

tequdev committed:
fix clang-format workflow

tequdev committed:
Merge remote-tracking branch 'upstream/dev' into jshooks-merge-20260225

RichardAH committed:
Merge pull request #697 from Xahau/jshooks-merge-20260225

JSHooks merge 20260225
